src/debugger.ts

```diff
@@ -1,212 +1,207 @@
 
+        // Configure debugger type
+        // On windows, use vs debugger if it's not mingw
+        config.type = 'cppdbg';
+        if (os.platform() == 'win32' && this.getPlat() != 'mingw') {
+            config.type = 'cppvsdbg';
         }
 
-        var customcfg = config.customDebugConfig;
-        for (let key in customcfg) {
-            debugConfig[key] = customcfg[key];
+        // Switch to lldb if needed
+        if (settings.debugConfigType == "codelldb") {
+            config.type = 'lldb';
+            config.stopOnEntry = config.stopAtEntry;
+            // CodeLLDB use program key for search a running procces
+            if (config.request == 'attach') {
+                config.stopOnEntry = false;
+                config.program = `${targetInformations.name}.exe`;
+            }
+        }
+        // Set MIMode for macos
+        if (os.platform() == 'darwin') {
+            config.MIMode = "lldb";
+            config.miDebuggerPath = "";
         }
-        // default type if not set yet
-        debugConfig.type = debugConfig.type || 'cppdbg';
 
-        // start debugging
-        if (debugConfig) {
-            await vscode.debug.startDebugging(vscode.workspace.workspaceFolders[0], debugConfig);
+        // Set MIMode for linux or windows/linux mingw
+        if (os.platform() == 'linux' || this.getPlat() == 'mingw') {
+            config.MIMode = "gdb";
+            config.miDebuggerPath = await findGdbPath();
         }
+
+        // Merge pretty printing
+        const setupCommands = {
+            setupCommands: [
+                {
+                    description: "Enable pretty-printing for gdb",
+                    text: "-enable-pretty-printing",
+                    ignoreFailures: true
+                }
+            ]
+        };
+        // Merge setupCommands
+        config = { ...config, ...setupCommands };
+
+        // Merge the custom debug config with actual config
+        config = { ...config, ...settings.customDebugConfig };
+
+        return config;
     }
 }
+
+export function initDebugger(context: vscode.ExtensionContext, option: Option) {
+    const cpptools = vscode.extensions.getExtension("ms-vscode.cpptools");
+    const codelldb = vscode.extensions.getExtension("vadimcn.vscode-lldb");
+
+    if (!cpptools && !codelldb) {
+        log.error("Neither CppTools or Code LLDB is installed");
+        vscode.window.showErrorMessage("Neither CppTools or Code LLDB is installed. Install one of this extension to debug");
+        return;
+    }
+
+    if (!codelldb && settings.debugConfigType == 'codelldb') {
+        vscode.window.showErrorMessage("Code LLDB is not installed. Install it first to use the debugger.");
+    }
+
+    // Activate the two extensions
+    cpptools?.activate();
+    codelldb?.activate()
+
+    const provider = new XmakeConfigurationProvider(option);
+    context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider('xmake', provider));
+}
```

Comment:
please add a new launchDebugger.ts to implement it, and please do not modify old debugger.ts

and you should add a new `onLaunchDebug()` command to bind launch debugger.

do not modify onDebug() now.

Reply:
If onLaunchDebug is stable in the future, I will merge the onDebug replacement with onLaunchDebug, but not just replace it now. This may cause some disruptive changes, as right now your patch is only for launch debugging support. Please do not modify the extra onDebug, or the debugging functionality on the toolbar.